using System;
using System.Collections.Generic;
using System.Dynamic;
using System.IO;
using Microsoft.VisualStudio.TestTools.UnitTesting;

namespace SimpleOpenTests.UnitTesting
{
    [TestClass]
    public class ReflectionAssertTest
    {
        [TestMethod]
        public void CultureTest()
        {
            // store current culture.
            var thread = System.Threading.Thread.CurrentThread;
            var currentCulture = thread.CurrentCulture;
            var currentUICulture = thread.CurrentUICulture;
            var newCultureName = string.Empty;
            Exception ex = null;

            // set "en-US".
            newCultureName = "en-US";
            thread.CurrentCulture = new System.Globalization.CultureInfo(newCultureName);
            thread.CurrentUICulture = new System.Globalization.CultureInfo(newCultureName);
            Console.WriteLine("CurrentCulture:{0} => {1}", currentCulture.Name, newCultureName);
            Console.WriteLine("CurrentUICulture:{0} => {1}", currentUICulture.Name, newCultureName);
            ex = new Exception();
            StringAssert.Contains(ex.Message, "was thrown");

            // set "ja-JP".
            newCultureName = "ja-JP";
            thread.CurrentCulture = new System.Globalization.CultureInfo(newCultureName);
            thread.CurrentUICulture = new System.Globalization.CultureInfo(newCultureName);
            Console.WriteLine("CurrentCulture:{0} => {1}", currentCulture.Name, newCultureName);
            Console.WriteLine("CurrentUICulture:{0} => {1}", currentUICulture.Name, newCultureName);
            ex = new Exception();
            Assert.IsFalse(ex.Message.Contains("was thrown"));

            // restore culture.
            thread.CurrentCulture = new System.Globalization.CultureInfo(currentCulture.Name);
            thread.CurrentUICulture = new System.Globalization.CultureInfo(currentUICulture.Name);
            Console.WriteLine("CurrentCulture:{0} => {1}", currentCulture.Name, newCultureName);
            Console.WriteLine("CurrentUICulture:{0} => {1}", currentUICulture.Name, newCultureName);
        }

        [TestMethod]
        public void HasProperty_2args_Test()
        {
            var dnActual = Directory.GetCurrentDirectory();
            var diActual = new DirectoryInfo(dnActual);
            var propName = "Name"; // exists.
            ReflectionAssert.HasProperty(diActual, propName);
        }

        [TestMethod] // do not use ExpectedExceptionAttribute().
        public void HasProperty_2args_ActualNullTest()
        {
            var dnActual = Directory.GetCurrentDirectory();
            var diActual = new DirectoryInfo(dnActual);
            var propName = "Name"; // exists.
            Exception exception = null;
            try
            {
                ReflectionAssert.HasProperty(null, propName); // actual = null
            }
            catch (ReflectionAssertException ex)
            {
                Console.WriteLine(ex.Message);
                exception = ex;
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                throw;
            }
            Assert.IsNotNull(exception);
            Assert.IsInstanceOfType(exception, typeof(ReflectionAssertException));
            StringAssert.Contains(exception.Message, "ReflectionAssert.HasProperty");
            Assert.IsNotNull(exception.InnerException);
            Assert.IsInstanceOfType(exception.InnerException, typeof(ArgumentNullException));
            StringAssert.Contains(exception.InnerException.Message, new ArgumentNullException().Message);
            StringAssert.Contains(exception.InnerException.Message, "actual");
        }

        [TestMethod] // do not use ExpectedExceptionAttribute().
        public void HasProperty_2args_PropertyNameNullTest()
        {
            var dnActual = Directory.GetCurrentDirectory();
            var diActual = new DirectoryInfo(dnActual);
            //var propName = "Name"; // exists.
            Exception exception = null;
            try
            {
                ReflectionAssert.HasProperty(diActual, (string)null); // propertyName = null
            }
            catch (ReflectionAssertException ex)
            {
                Console.WriteLine(ex.Message);
                exception = ex;
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                throw;
            }
            Assert.IsNotNull(exception);
            Assert.IsInstanceOfType(exception, typeof(ReflectionAssertException));
            StringAssert.Contains(exception.Message, "ReflectionAssert.HasProperty");
            Assert.IsNotNull(exception.InnerException);
            Assert.IsInstanceOfType(exception.InnerException, typeof(ArgumentNullException));
            StringAssert.Contains(exception.InnerException.Message, new ArgumentNullException().Message);
            StringAssert.Contains(exception.InnerException.Message, "propertyName");
        }

        [TestMethod] // do not use ExpectedExceptionAttribute().
        public void HasProperty_2args_PropertyNameEmptyTest()
        {
            var dnActual = Directory.GetCurrentDirectory();
            var diActual = new DirectoryInfo(dnActual);
            //var propName = "Name"; // exists.
            Exception exception = null;
            try
            {
                ReflectionAssert.HasProperty(diActual, string.Empty); // propertyName = string.Empty
            }
            catch (ReflectionAssertException ex)
            {
                Console.WriteLine(ex.Message);
                exception = ex;
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                throw;
            }
            Assert.IsNotNull(exception);
            Assert.IsInstanceOfType(exception, typeof(ReflectionAssertException));
            StringAssert.Contains(exception.Message, "ReflectionAssert.HasProperty");
            Assert.IsNotNull(exception.InnerException);
            Assert.IsInstanceOfType(exception.InnerException, typeof(ArgumentException));
            StringAssert.Contains(exception.InnerException.Message, new ArgumentException().Message);
            StringAssert.Contains(exception.InnerException.Message, "propertyName");
        }

        [TestMethod] // do not use ExpectedExceptionAttribute().
        public void HasProperty_2args_PropertyNameNotExistTest()
        {
            var dnActual = Directory.GetCurrentDirectory();
            var diActual = new DirectoryInfo(dnActual);
            var propName = "NoName"; // not exist.
            Exception exception = null;
            try
            {
                ReflectionAssert.HasProperty(diActual, propName);
            }
            catch (ReflectionAssertException ex)
            {
                Console.WriteLine(ex.Message);
                exception = ex;
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                throw;
            }
            Assert.IsNotNull(exception);
            Assert.IsInstanceOfType(exception, typeof(ReflectionAssertException));
            StringAssert.Contains(exception.Message, "ReflectionAssert.HasProperty");
            Assert.IsNotNull(exception.InnerException);
            Assert.IsInstanceOfType(exception.InnerException, typeof(ArgumentException));
            //StringAssert.Contains(exception.InnerException.Message, new ArgumentException().Message);
            StringAssert.Contains(exception.InnerException.Message, propName);
        }

        // *** dynamic is not supported.

        [TestMethod]
        [ExpectedException(typeof(Microsoft.CSharp.RuntimeBinder.RuntimeBinderException))]
        public void HasProperty_2args_dynamic_Test()
        {
            var dnActual = Directory.GetCurrentDirectory();
            var diActual = new DirectoryInfo(dnActual);
            var propName = "Name"; // exists.
            dynamic dynamicActual = new ExpandoObject();
            dynamicActual.Name = diActual.Name;
            ReflectionAssert.HasProperty(dynamicActual, propName);
        }

        [TestMethod]
        public void HasProperty_2args_msg_Test()
        {
            var dnActual = Directory.GetCurrentDirectory();
            var diActual = new DirectoryInfo(dnActual);
            var propName = "Name"; // exists.
            var message = "message";
            ReflectionAssert.HasProperty(diActual, propName, message);
        }

        [TestMethod] // do not use ExpectedExceptionAttribute().
        public void HasProperty_2args_msg_ActualNullTest()
        {
            var dnActual = Directory.GetCurrentDirectory();
            var diActual = new DirectoryInfo(dnActual);
            var propName = "Name"; // exists.
            var message = "AdditionalMessage.";
            Exception exception = null;
            try
            {
                ReflectionAssert.HasProperty(null, propName, message); // actual = null.
            }
            catch (ReflectionAssertException ex)
            {
                Console.WriteLine(ex.Message);
                exception = ex;
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                throw;
            }
            Assert.IsNotNull(exception);
            Assert.IsInstanceOfType(exception, typeof(ReflectionAssertException));
            StringAssert.Contains(exception.Message, "ReflectionAssert.HasProperty");
            StringAssert.Contains(exception.Message, message);
            Assert.IsNotNull(exception.InnerException);
            Assert.IsInstanceOfType(exception.InnerException, typeof(ArgumentNullException));
            StringAssert.Contains(exception.InnerException.Message, new ArgumentNullException().Message);
            StringAssert.Contains(exception.InnerException.Message, "actual");
        }

        [TestMethod] // do not use ExpectedExceptionAttribute().
        public void HasProperty_2args_msg_PropertyNameNullTest()
        {
            var dnActual = Directory.GetCurrentDirectory();
            var diActual = new DirectoryInfo(dnActual);
            //var propName = "Name"; // exists.
            var message = "AdditionalMessage.";
            Exception exception = null;
            try
            {
                ReflectionAssert.HasProperty(diActual, (string)null, message);
            }
            catch (ReflectionAssertException ex)
            {
                Console.WriteLine(ex.Message);
                exception = ex;
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                throw;
            }
            Assert.IsNotNull(exception);
            Assert.IsInstanceOfType(exception, typeof(ReflectionAssertException));
            StringAssert.Contains(exception.Message, "ReflectionAssert.HasProperty");
            StringAssert.Contains(exception.Message, message);
            Assert.IsNotNull(exception.InnerException);
            Assert.IsInstanceOfType(exception.InnerException, typeof(ArgumentNullException));
            StringAssert.Contains(exception.InnerException.Message, new ArgumentNullException().Message);
            StringAssert.Contains(exception.InnerException.Message, "propertyName");
        }

        [TestMethod] // do not use ExpectedExceptionAttribute().
        public void HasProperty_2args_msg_PropertyNameEmptyTest()
        {
            var dnActual = Directory.GetCurrentDirectory();
            var diActual = new DirectoryInfo(dnActual);
            //var propName = "Name"; // exists.
            var message = "AdditionalMessage.";
            Exception exception = null;
            try
            {
                ReflectionAssert.HasProperty(diActual, string.Empty, message);
            }
            catch (ReflectionAssertException ex)
            {
                Console.WriteLine(ex.Message);
                exception = ex;
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                throw;
            }
            Assert.IsNotNull(exception);
            Assert.IsInstanceOfType(exception, typeof(ReflectionAssertException));
            StringAssert.Contains(exception.Message, "ReflectionAssert.HasProperty");
            StringAssert.Contains(exception.Message, message);
            Assert.IsNotNull(exception.InnerException);
            Assert.IsInstanceOfType(exception.InnerException, typeof(ArgumentException));
            StringAssert.Contains(exception.InnerException.Message, new ArgumentException().Message);
            StringAssert.Contains(exception.InnerException.Message, "propertyName");
        }

        [TestMethod] // do not use ExpectedExceptionAttribute().
        public void HasProperty_2args_msg_PropertyNameNotExistTest()
        {
            var dnActual = Directory.GetCurrentDirectory();
            var diActual = new DirectoryInfo(dnActual);
            var propName = "NoName"; // not exist.
            var message = "AdditionalMessage.";
            Exception exception = null;
            try
            {
                ReflectionAssert.HasProperty(diActual, propName, message);
            }
            catch (ReflectionAssertException ex)
            {
                Console.WriteLine(ex.Message);
                exception = ex;
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                throw;
            }
            Assert.IsNotNull(exception);
            Assert.IsInstanceOfType(exception, typeof(ReflectionAssertException));
            StringAssert.Contains(exception.Message, "ReflectionAssert.HasProperty");
            StringAssert.Contains(exception.Message, message);
            Assert.IsNotNull(exception.InnerException);
            Assert.IsInstanceOfType(exception.InnerException, typeof(ArgumentException));
            //StringAssert.Contains(exception.InnerException.Message, new ArgumentException().Message);
            StringAssert.Contains(exception.InnerException.Message, propName);
        }

        [TestMethod]
        public void HasProperty_3args_Test()
        {
            var dnActual = Directory.GetCurrentDirectory();
            var diActual = new DirectoryInfo(dnActual);
            var propName = "Name"; // exists.
            var propType = typeof(string); // string.
            ReflectionAssert.HasProperty(diActual, propName, propType);
        }

        [TestMethod] // do not use ExpectedExceptionAttribute().
        public void HasProperty_3args_ActualNullTest()
        {
            var dnActual = Directory.GetCurrentDirectory();
            var diActual = new DirectoryInfo(dnActual);
            var propName = "Name"; // exists.
            var propType = typeof(string); // string.
            Exception exception = null;
            try
            {
                ReflectionAssert.HasProperty(null, propName, propType); // actual = null
            }
            catch (ReflectionAssertException ex)
            {
                Console.WriteLine(ex.Message);
                exception = ex;
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                throw;
            }
            Assert.IsNotNull(exception);
            Assert.IsInstanceOfType(exception, typeof(ReflectionAssertException));
            StringAssert.Contains(exception.Message, "ReflectionAssert.HasProperty");
            Assert.IsNotNull(exception.InnerException);
            Assert.IsInstanceOfType(exception.InnerException, typeof(ArgumentNullException));
            StringAssert.Contains(exception.InnerException.Message, new ArgumentNullException().Message);
            StringAssert.Contains(exception.InnerException.Message, "actual");
        }

        [TestMethod] // do not use ExpectedExceptionAttribute().
        public void HasProperty_3args_PropertyNameNullTest()
        {
            var dnActual = Directory.GetCurrentDirectory();
            var diActual = new DirectoryInfo(dnActual);
            //var propName = "Name"; // exists.
            var propType = typeof(string); // string.
            Exception exception = null;
            try
            {
                ReflectionAssert.HasProperty(diActual, (string)null, propType); // propertyName = null
            }
            catch (ReflectionAssertException ex)
            {
                Console.WriteLine(ex.Message);
                exception = ex;
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                throw;
            }
            Assert.IsNotNull(exception);
            Assert.IsInstanceOfType(exception, typeof(ReflectionAssertException));
            StringAssert.Contains(exception.Message, "ReflectionAssert.HasProperty");
            Assert.IsNotNull(exception.InnerException);
            Assert.IsInstanceOfType(exception.InnerException, typeof(ArgumentNullException));
            StringAssert.Contains(exception.InnerException.Message, new ArgumentNullException().Message);
            StringAssert.Contains(exception.InnerException.Message, "propertyName");
        }

        [TestMethod] // do not use ExpectedExceptionAttribute().
        public void HasProperty_3args_PropertyNameEmptyTest()
        {
            var dnActual = Directory.GetCurrentDirectory();
            var diActual = new DirectoryInfo(dnActual);
            //var propName = "Name"; // exists.
            var propType = typeof(string); // string.
            Exception exception = null;
            try
            {
                ReflectionAssert.HasProperty(diActual, string.Empty, propType); // propertyName = string.Empty
            }
            catch (ReflectionAssertException ex)
            {
                Console.WriteLine(ex.Message);
                exception = ex;
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                throw;
            }
            Assert.IsNotNull(exception);
            Assert.IsInstanceOfType(exception, typeof(ReflectionAssertException));
            StringAssert.Contains(exception.Message, "ReflectionAssert.HasProperty");
            Assert.IsNotNull(exception.InnerException);
            Assert.IsInstanceOfType(exception.InnerException, typeof(ArgumentException));
            StringAssert.Contains(exception.InnerException.Message, new ArgumentException().Message);
            StringAssert.Contains(exception.InnerException.Message, "propertyName");
        }

        [TestMethod] // do not use ExpectedExceptionAttribute().
        public void HasProperty_3args_PropertyNameNotExistTest()
        {
            var dnActual = Directory.GetCurrentDirectory();
            var diActual = new DirectoryInfo(dnActual);
            var propName = "NoName"; // not exist.
            var propType = typeof(string); // string.
            Exception exception = null;
            try
            {
                ReflectionAssert.HasProperty(diActual, propName, propType);
            }
            catch (ReflectionAssertException ex)
            {
                Console.WriteLine(ex.Message);
                exception = ex;
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                throw;
            }
            Assert.IsNotNull(exception);
            Assert.IsInstanceOfType(exception, typeof(ReflectionAssertException));
            StringAssert.Contains(exception.Message, "ReflectionAssert.HasProperty");
            Assert.IsNotNull(exception.InnerException);
            Assert.IsInstanceOfType(exception.InnerException, typeof(ArgumentException));
            //StringAssert.Contains(exception.InnerException.Message, new ArgumentException().Message);
            StringAssert.Contains(exception.InnerException.Message, propName);
        }

        [TestMethod] // do not use ExpectedExceptionAttribute().
        public void HasProperty_3args_PropertyTypeNullTest()
        {
            var dnActual = Directory.GetCurrentDirectory();
            var diActual = new DirectoryInfo(dnActual);
            var propName = "Name"; // exists.
            //var propType = typeof(string); // string.
            Exception exception = null;
            try
            {
                ReflectionAssert.HasProperty(diActual, propName, (Type)null); // propertyType = null
            }
            catch (ReflectionAssertException ex)
            {
                Console.WriteLine(ex.Message);
                exception = ex;
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                throw;
            }
            Assert.IsNotNull(exception);
            Assert.IsInstanceOfType(exception, typeof(ReflectionAssertException));
            StringAssert.Contains(exception.Message, "ReflectionAssert.HasProperty");
            Assert.IsNotNull(exception.InnerException);
            Assert.IsInstanceOfType(exception.InnerException, typeof(ArgumentNullException));
            StringAssert.Contains(exception.InnerException.Message, new ArgumentNullException().Message);
            StringAssert.Contains(exception.InnerException.Message, "propertyType");
        }

        [TestMethod] // do not use ExpectedExceptionAttribute().
        public void HasProperty_3args_PropertyTypeMismatchTypeTest()
        {
            var dnActual = Directory.GetCurrentDirectory();
            var diActual = new DirectoryInfo(dnActual);
            var propName = "Name"; // exists.
            //var propType = typeof(string); // string.
            Exception exception = null;
            try
            {
                ReflectionAssert.HasProperty(diActual, propName, typeof(int)); // propertyType = typeof(int)
            }
            catch (ReflectionAssertException ex)
            {
                Console.WriteLine(ex.Message);
                exception = ex;
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                throw;
            }
            Assert.IsNotNull(exception);
            Assert.IsInstanceOfType(exception, typeof(ReflectionAssertException));
            StringAssert.Contains(exception.Message, "ReflectionAssert.HasProperty");
            Assert.IsNotNull(exception.InnerException);
            Assert.IsInstanceOfType(exception.InnerException, typeof(ArgumentException));
            StringAssert.Contains(exception.InnerException.Message, new InvalidCastException().Message);
            StringAssert.Contains(exception.InnerException.Message, "propertyType");
        }

        [TestMethod]
        public void HasProperty_3args_msg_Test()
        {
            var dnActual = Directory.GetCurrentDirectory();
            var diActual = new DirectoryInfo(dnActual);
            var propName = "Name"; // exists.
            var propType = typeof(string); // string.
            var message = "message";
            ReflectionAssert.HasProperty(diActual, propName, propType, message);
        }

        [TestMethod] // do not use ExpectedExceptionAttribute().
        public void HasProperty_3args_msg_ActualNullTest()
        {
            var dnActual = Directory.GetCurrentDirectory();
            var diActual = new DirectoryInfo(dnActual);
            var propName = "Name"; // exists.
            var propType = typeof(string); // string.
            var message = "message";
            Exception exception = null;
            try
            {
                ReflectionAssert.HasProperty(null, propName, propType, message); // actual = null
            }
            catch (ReflectionAssertException ex)
            {
                Console.WriteLine(ex.Message);
                exception = ex;
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                throw;
            }
            Assert.IsNotNull(exception);
            Assert.IsInstanceOfType(exception, typeof(ReflectionAssertException));
            StringAssert.Contains(exception.Message, "ReflectionAssert.HasProperty");
            StringAssert.Contains(exception.Message, message);
            Assert.IsNotNull(exception.InnerException);
            Assert.IsInstanceOfType(exception.InnerException, typeof(ArgumentNullException));
            StringAssert.Contains(exception.InnerException.Message, new ArgumentNullException().Message);
            StringAssert.Contains(exception.InnerException.Message, "actual");
        }

        [TestMethod] // do not use ExpectedExceptionAttribute().
        public void HasProperty_3args_msg_PropertyNameNullTest()
        {
            var dnActual = Directory.GetCurrentDirectory();
            var diActual = new DirectoryInfo(dnActual);
            //var propName = "Name"; // exists.
            var propType = typeof(string); // string.
            var message = "message";
            Exception exception = null;
            try
            {
                ReflectionAssert.HasProperty(diActual, (string)null, propType, message); // propertyName = null
            }
            catch (ReflectionAssertException ex)
            {
                Console.WriteLine(ex.Message);
                exception = ex;
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                throw;
            }
            Assert.IsNotNull(exception);
            Assert.IsInstanceOfType(exception, typeof(ReflectionAssertException));
            StringAssert.Contains(exception.Message, "ReflectionAssert.HasProperty");
            StringAssert.Contains(exception.Message, message);
            Assert.IsNotNull(exception.InnerException);
            Assert.IsInstanceOfType(exception.InnerException, typeof(ArgumentNullException));
            StringAssert.Contains(exception.InnerException.Message, new ArgumentNullException().Message);
            StringAssert.Contains(exception.InnerException.Message, "propertyName");
        }

        [TestMethod] // do not use ExpectedExceptionAttribute().
        public void HasProperty_3args_msg_PropertyNameEmptyTest()
        {
            var dnActual = Directory.GetCurrentDirectory();
            var diActual = new DirectoryInfo(dnActual);
            //var propName = "Name"; // exists.
            var propType = typeof(string); // string.
            var message = "message";
            Exception exception = null;
            try
            {
                ReflectionAssert.HasProperty(diActual, string.Empty, propType, message); // propertyName = string.Empty
            }
            catch (ReflectionAssertException ex)
            {
                Console.WriteLine(ex.Message);
                exception = ex;
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                throw;
            }
            Assert.IsNotNull(exception);
            Assert.IsInstanceOfType(exception, typeof(ReflectionAssertException));
            StringAssert.Contains(exception.Message, "ReflectionAssert.HasProperty");
            StringAssert.Contains(exception.Message, message);
            Assert.IsNotNull(exception.InnerException);
            Assert.IsInstanceOfType(exception.InnerException, typeof(ArgumentException));
            StringAssert.Contains(exception.InnerException.Message, new ArgumentException().Message);
            StringAssert.Contains(exception.InnerException.Message, "propertyName");
        }

        [TestMethod] // do not use ExpectedExceptionAttribute().
        public void HasProperty_3args_msg_PropertyNameNotExistTest()
        {
            var dnActual = Directory.GetCurrentDirectory();
            var diActual = new DirectoryInfo(dnActual);
            var propName = "NoName"; // not exist.
            var propType = typeof(string); // string.
            var message = "message";
            Exception exception = null;
            try
            {
                ReflectionAssert.HasProperty(diActual, propName, propType, message);
            }
            catch (ReflectionAssertException ex)
            {
                Console.WriteLine(ex.Message);
                exception = ex;
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                throw;
            }
            Assert.IsNotNull(exception);
            Assert.IsInstanceOfType(exception, typeof(ReflectionAssertException));
            StringAssert.Contains(exception.Message, "ReflectionAssert.HasProperty");
            StringAssert.Contains(exception.Message, message);
            Assert.IsNotNull(exception.InnerException);
            Assert.IsInstanceOfType(exception.InnerException, typeof(ArgumentException));
            //StringAssert.Contains(exception.InnerException.Message, new ArgumentException().Message);
            StringAssert.Contains(exception.InnerException.Message, propName);
        }

        [TestMethod] // do not use ExpectedExceptionAttribute().
        public void HasProperty_3args_msg_PropertyTypeNullTest()
        {
            var dnActual = Directory.GetCurrentDirectory();
            var diActual = new DirectoryInfo(dnActual);
            var propName = "Name"; // exists.
            //var propType = typeof(string); // string.
            var message = "message";
            Exception exception = null;
            try
            {
                ReflectionAssert.HasProperty(diActual, propName, (Type)null, message); // propertyType = null
            }
            catch (ReflectionAssertException ex)
            {
                Console.WriteLine(ex.Message);
                exception = ex;
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                throw;
            }
            Assert.IsNotNull(exception);
            Assert.IsInstanceOfType(exception, typeof(ReflectionAssertException));
            StringAssert.Contains(exception.Message, "ReflectionAssert.HasProperty");
            StringAssert.Contains(exception.Message, message);
            Assert.IsNotNull(exception.InnerException);
            Assert.IsInstanceOfType(exception.InnerException, typeof(ArgumentNullException));
            StringAssert.Contains(exception.InnerException.Message, new ArgumentNullException().Message);
            StringAssert.Contains(exception.InnerException.Message, "propertyType");
        }

        [TestMethod] // do not use ExpectedExceptionAttribute().
        public void HasProperty_3args_msg_PropertyTypeMismatchTypeTest()
        {
            var dnActual = Directory.GetCurrentDirectory();
            var diActual = new DirectoryInfo(dnActual);
            var propName = "Name"; // exists.
            //var propType = typeof(string); // string.
            var message = "message";
            Exception exception = null;
            try
            {
                ReflectionAssert.HasProperty(diActual, propName, typeof(int), message); // propertyType = typeof(int)
            }
            catch (ReflectionAssertException ex)
            {
                Console.WriteLine(ex.Message);
                exception = ex;
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                throw;
            }
            Assert.IsNotNull(exception);
            Assert.IsInstanceOfType(exception, typeof(ReflectionAssertException));
            StringAssert.Contains(exception.Message, "ReflectionAssert.HasProperty");
            StringAssert.Contains(exception.Message, message);
            Assert.IsNotNull(exception.InnerException);
            Assert.IsInstanceOfType(exception.InnerException, typeof(ArgumentException));
            StringAssert.Contains(exception.InnerException.Message, new InvalidCastException().Message);
            StringAssert.Contains(exception.InnerException.Message, "propertyType");
        }

        [TestMethod]
        public void HasProperties_2args_Test()
        {
            var dnActual = Directory.GetCurrentDirectory();
            var diActual = new DirectoryInfo(dnActual);
            var propNames = new[] { "Name", "FullName" }; // exist.
            ReflectionAssert.HasProperties(diActual, propNames);
        }

        [TestMethod] // do not use ExpectedExceptionAttribute().
        public void HasProperties_2args_ActualNullTest()
        {
            var dnActual = Directory.GetCurrentDirectory();
            var diActual = new DirectoryInfo(dnActual);
            var propNames = new[] { "Name", "FullName" }; // exist.
            Exception exception = null;
            try
            {
                ReflectionAssert.HasProperties(null, propNames); // actual = null
            }
            catch (ReflectionAssertException ex)
            {
                Console.WriteLine(ex.Message);
                exception = ex;
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                throw;
            }
            Assert.IsNotNull(exception);
            Assert.IsInstanceOfType(exception, typeof(ReflectionAssertException));
            StringAssert.Contains(exception.Message, "ReflectionAssert.HasProperties");
            Assert.IsNotNull(exception.InnerException);
            Assert.IsInstanceOfType(exception.InnerException, typeof(ArgumentNullException));
            StringAssert.Contains(exception.InnerException.Message, new ArgumentNullException().Message);
            StringAssert.Contains(exception.InnerException.Message, "actual");
        }

        [TestMethod] // do not use ExpectedExceptionAttribute().
        public void HasProperties_2args_PropertyNamesNullTest()
        {
            var dnActual = Directory.GetCurrentDirectory();
            var diActual = new DirectoryInfo(dnActual);
            //var propNames = new[] { "Name", "FullName" }; // exist.
            Exception exception = null;
            try
            {
                ReflectionAssert.HasProperties(diActual, (IEnumerable<string>)null); // propertyNames = null
            }
            catch (ReflectionAssertException ex)
            {
                Console.WriteLine(ex.Message);
                exception = ex;
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                throw;
            }
            Assert.IsNotNull(exception);
            Assert.IsInstanceOfType(exception, typeof(ReflectionAssertException));
            StringAssert.Contains(exception.Message, "ReflectionAssert.HasProperties");
            Assert.IsNotNull(exception.InnerException);
            Assert.IsInstanceOfType(exception.InnerException, typeof(ArgumentNullException));
            StringAssert.Contains(exception.InnerException.Message, new ArgumentNullException().Message);
            StringAssert.Contains(exception.InnerException.Message, "propertyNames");
        }

        [TestMethod] // do not use ExpectedExceptionAttribute().
        public void HasProperties_2args_PropertyNamesContainsNullTest()
        {
            var dnActual = Directory.GetCurrentDirectory();
            var diActual = new DirectoryInfo(dnActual);
            var propNames = new[] { "Name", "FullName", null }; // contains null.
            Exception exception = null;
            try
            {
                ReflectionAssert.HasProperties(diActual, propNames); // propertyNames contains null
            }
            catch (ReflectionAssertException ex)
            {
                Console.WriteLine(ex.Message);
                exception = ex;
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                throw;
            }
            Assert.IsNotNull(exception);
            Assert.IsInstanceOfType(exception, typeof(ReflectionAssertException));
            StringAssert.Contains(exception.Message, "ReflectionAssert.HasProperties");
            Assert.IsNotNull(exception.InnerException);
            Assert.IsInstanceOfType(exception.InnerException, typeof(ArgumentNullException));
            StringAssert.Contains(exception.InnerException.Message, new ArgumentNullException().Message);
            StringAssert.Contains(exception.InnerException.Message, "propertyNames");
        }

        [TestMethod] // do not use ExpectedExceptionAttribute().
        public void HasProperties_2args_PropertyNamesContainsEmptyTest()
        {
            var dnActual = Directory.GetCurrentDirectory();
            var diActual = new DirectoryInfo(dnActual);
            var propNames = new[] { "Name", "FullName", string.Empty }; // contains string.Empty.
            Exception exception = null;
            try
            {
                ReflectionAssert.HasProperties(diActual, propNames); // propertyNames contains string.Empty
            }
            catch (ReflectionAssertException ex)
            {
                Console.WriteLine(ex.Message);
                exception = ex;
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                throw;
            }
            Assert.IsNotNull(exception);
            Assert.IsInstanceOfType(exception, typeof(ReflectionAssertException));
            StringAssert.Contains(exception.Message, "ReflectionAssert.HasProperties");
            Assert.IsNotNull(exception.InnerException);
            Assert.IsInstanceOfType(exception.InnerException, typeof(ArgumentException));
            StringAssert.Contains(exception.InnerException.Message, new ArgumentException().Message);
            StringAssert.Contains(exception.InnerException.Message, "propertyNames");
        }

        [TestMethod] // do not use ExpectedExceptionAttribute().
        public void HasProperties_2args_PropertyNamesContainsNotExistTest()
        {
            var dnActual = Directory.GetCurrentDirectory();
            var diActual = new DirectoryInfo(dnActual);
            var propNames = new[] { "Name", "FullName", "NoName" }; // contains no exist.
            Exception exception = null;
            try
            {
                ReflectionAssert.HasProperties(diActual, propNames);
            }
            catch (ReflectionAssertException ex)
            {
                Console.WriteLine(ex.Message);
                exception = ex;
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                throw;
            }
            Assert.IsNotNull(exception);
            Assert.IsInstanceOfType(exception, typeof(ReflectionAssertException));
            StringAssert.Contains(exception.Message, "ReflectionAssert.HasProperties");
            Assert.IsNotNull(exception.InnerException);
            Assert.IsInstanceOfType(exception.InnerException, typeof(ArgumentException));
            //StringAssert.Contains(exception.InnerException.Message, new ArgumentException().Message);
            StringAssert.Contains(exception.InnerException.Message, "NoName");
        }

        [TestMethod]
        public void HasProperties_2args_msg_Test()
        {
            var dnActual = Directory.GetCurrentDirectory();
            var diActual = new DirectoryInfo(dnActual);
            var propNames = new[] { "Name", "FullName" }; // exist.
            var message = "message";
            ReflectionAssert.HasProperties(diActual, propNames, message);
        }

        [TestMethod] // do not use ExpectedExceptionAttribute().
        public void HasProperties_2args_msg_ActualNullTest()
        {
            var dnActual = Directory.GetCurrentDirectory();
            var diActual = new DirectoryInfo(dnActual);
            var propNames = new[] { "Name", "FullName" }; // exist.
            var message = "message";
            Exception exception = null;
            try
            {
                ReflectionAssert.HasProperties(null, propNames, message); // actual = null
            }
            catch (ReflectionAssertException ex)
            {
                Console.WriteLine(ex.Message);
                exception = ex;
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                throw;
            }
            Assert.IsNotNull(exception);
            Assert.IsInstanceOfType(exception, typeof(ReflectionAssertException));
            StringAssert.Contains(exception.Message, "ReflectionAssert.HasProperties");
            StringAssert.Contains(exception.Message, message);
            Assert.IsNotNull(exception.InnerException);
            Assert.IsInstanceOfType(exception.InnerException, typeof(ArgumentNullException));
            StringAssert.Contains(exception.InnerException.Message, new ArgumentNullException().Message);
            StringAssert.Contains(exception.InnerException.Message, "actual");
        }

        [TestMethod] // do not use ExpectedExceptionAttribute().
        public void HasProperties_2args_msg_PropertyNamesNullTest()
        {
            var dnActual = Directory.GetCurrentDirectory();
            var diActual = new DirectoryInfo(dnActual);
            //var propNames = new[] { "Name", "FullName" }; // exist.
            var message = "message";
            Exception exception = null;
            try
            {
                ReflectionAssert.HasProperties(diActual, (IEnumerable<string>)null, message); // propertyNames = null
            }
            catch (ReflectionAssertException ex)
            {
                Console.WriteLine(ex.Message);
                exception = ex;
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                throw;
            }
            Assert.IsNotNull(exception);
            Assert.IsInstanceOfType(exception, typeof(ReflectionAssertException));
            StringAssert.Contains(exception.Message, "ReflectionAssert.HasProperties");
            StringAssert.Contains(exception.Message, message);
            Assert.IsNotNull(exception.InnerException);
            Assert.IsInstanceOfType(exception.InnerException, typeof(ArgumentNullException));
            StringAssert.Contains(exception.InnerException.Message, new ArgumentNullException().Message);
            StringAssert.Contains(exception.InnerException.Message, "propertyNames");
        }

        [TestMethod] // do not use ExpectedExceptionAttribute().
        public void HasProperties_2args_msg_PropertyNamesContainsNullTest()
        {
            var dnActual = Directory.GetCurrentDirectory();
            var diActual = new DirectoryInfo(dnActual);
            var propNames = new[] { "Name", "FullName", null }; // contains null.
            var message = "message";
            Exception exception = null;
            try
            {
                ReflectionAssert.HasProperties(diActual, (IEnumerable<string>)null, message); // propertyNames contains null
            }
            catch (ReflectionAssertException ex)
            {
                Console.WriteLine(ex.Message);
                exception = ex;
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                throw;
            }
            Assert.IsNotNull(exception);
            Assert.IsInstanceOfType(exception, typeof(ReflectionAssertException));
            StringAssert.Contains(exception.Message, "ReflectionAssert.HasProperties");
            StringAssert.Contains(exception.Message, message);
            Assert.IsNotNull(exception.InnerException);
            Assert.IsInstanceOfType(exception.InnerException, typeof(ArgumentNullException));
            StringAssert.Contains(exception.InnerException.Message, new ArgumentNullException().Message);
            StringAssert.Contains(exception.InnerException.Message, "propertyNames");
        }

        [TestMethod] // do not use ExpectedExceptionAttribute().
        public void HasProperties_2args_msg_PropertyNamesContainsEmptyTest()
        {
            var dnActual = Directory.GetCurrentDirectory();
            var diActual = new DirectoryInfo(dnActual);
            var propNames = new[] { "Name", "FullName", string.Empty }; // contains string.Empty.
            var message = "message";
            Exception exception = null;
            try
            {
                ReflectionAssert.HasProperties(diActual, propNames, message); // propertyNames contains string.Empty
            }
            catch (ReflectionAssertException ex)
            {
                Console.WriteLine(ex.Message);
                exception = ex;
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                throw;
            }
            Assert.IsNotNull(exception);
            Assert.IsInstanceOfType(exception, typeof(ReflectionAssertException));
            StringAssert.Contains(exception.Message, "ReflectionAssert.HasProperties");
            StringAssert.Contains(exception.Message, message);
            Assert.IsNotNull(exception.InnerException);
            Assert.IsInstanceOfType(exception.InnerException, typeof(ArgumentException));
            StringAssert.Contains(exception.InnerException.Message, new ArgumentException().Message);
            StringAssert.Contains(exception.InnerException.Message, "propertyNames");
        }

        [TestMethod] // do not use ExpectedExceptionAttribute().
        public void HasProperties_2args_msg_PropertyNamesContainsNotExistTest()
        {
            var dnActual = Directory.GetCurrentDirectory();
            var diActual = new DirectoryInfo(dnActual);
            var propNames = new[] { "Name", "FullName", "NoName" }; // contains no exist.
            var message = "message";
            Exception exception = null;
            try
            {
                ReflectionAssert.HasProperties(diActual, propNames, message);
            }
            catch (ReflectionAssertException ex)
            {
                Console.WriteLine(ex.Message);
                exception = ex;
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                throw;
            }
            Assert.IsNotNull(exception);
            Assert.IsInstanceOfType(exception, typeof(ReflectionAssertException));
            StringAssert.Contains(exception.Message, "ReflectionAssert.HasProperties");
            StringAssert.Contains(exception.Message, message);
            Assert.IsNotNull(exception.InnerException);
            Assert.IsInstanceOfType(exception.InnerException, typeof(ArgumentException));
            //StringAssert.Contains(exception.InnerException.Message, new ArgumentException().Message);
            StringAssert.Contains(exception.InnerException.Message, "NoName");
        }

        [TestMethod]
        public void HasProperties_3args_Test()
        {
            var dnActual = Directory.GetCurrentDirectory();
            var diActual = new DirectoryInfo(dnActual);
            var propNames = new[] { "Name", "FullName" }; // exist.
            var propTypes = new[] { typeof(string), typeof(string) };
            ReflectionAssert.HasProperties(diActual, propNames);
        }

        [TestMethod] // do not use ExpectedExceptionAttribute().
        public void HasProperties_3args_ActualNullTest()
        {
            var dnActual = Directory.GetCurrentDirectory();
            var diActual = new DirectoryInfo(dnActual);
            var propNames = new[] { "Name", "FullName" }; // exist.
            var propTypes = new[] { typeof(string), typeof(string) };
            Exception exception = null;
            try
            {
                ReflectionAssert.HasProperties(null, propNames, propTypes); // actual = null
            }
            catch (ReflectionAssertException ex)
            {
                Console.WriteLine(ex.Message);
                exception = ex;
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                throw;
            }
            Assert.IsNotNull(exception);
            Assert.IsInstanceOfType(exception, typeof(ReflectionAssertException));
            StringAssert.Contains(exception.Message, "ReflectionAssert.HasProperties");
            Assert.IsNotNull(exception.InnerException);
            Assert.IsInstanceOfType(exception.InnerException, typeof(ArgumentNullException));
            StringAssert.Contains(exception.InnerException.Message, new ArgumentNullException().Message);
            StringAssert.Contains(exception.InnerException.Message, "actual");
        }

        [TestMethod] // do not use ExpectedExceptionAttribute().
        public void HasProperties_3args_PropertyNamesNullTest()
        {
            var dnActual = Directory.GetCurrentDirectory();
            var diActual = new DirectoryInfo(dnActual);
            var propNames = new[] { "Name", "FullName" }; // exist.
            var propTypes = new[] { typeof(string), typeof(string) };
            Exception exception = null;
            try
            {
                ReflectionAssert.HasProperties(diActual, (IEnumerable<string>)null, propTypes); // propertyNames = null
            }
            catch (ReflectionAssertException ex)
            {
                Console.WriteLine(ex.Message);
                exception = ex;
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                throw;
            }
            Assert.IsNotNull(exception);
            Assert.IsInstanceOfType(exception, typeof(ReflectionAssertException));
            StringAssert.Contains(exception.Message, "ReflectionAssert.HasProperties");
            Assert.IsNotNull(exception.InnerException);
            Assert.IsInstanceOfType(exception.InnerException, typeof(ArgumentNullException));
            StringAssert.Contains(exception.InnerException.Message, new ArgumentNullException().Message);
            StringAssert.Contains(exception.InnerException.Message, "propertyNames");
        }

        [TestMethod] // do not use ExpectedExceptionAttribute().
        public void HasProperties_3args_PropertyNamesContainsNullTest()
        {
            var dnActual = Directory.GetCurrentDirectory();
            var diActual = new DirectoryInfo(dnActual);
            var propNames = new[] { "Name", "FullName", null }; // contains null.
            var propTypes = new[] { typeof(string), typeof(string), typeof(string) };
            Exception exception = null;
            try
            {
                ReflectionAssert.HasProperties(diActual, propNames, propTypes); // propertyNames contains null
            }
            catch (ReflectionAssertException ex)
            {
                Console.WriteLine(ex.Message);
                exception = ex;
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                throw;
            }
            Assert.IsNotNull(exception);
            Assert.IsInstanceOfType(exception, typeof(ReflectionAssertException));
            StringAssert.Contains(exception.Message, "ReflectionAssert.HasProperties");
            Assert.IsNotNull(exception.InnerException);
            Assert.IsInstanceOfType(exception.InnerException, typeof(ArgumentNullException));
            StringAssert.Contains(exception.InnerException.Message, new ArgumentNullException().Message);
            StringAssert.Contains(exception.InnerException.Message, "propertyNames");
        }

        [TestMethod] // do not use ExpectedExceptionAttribute().
        public void HasProperties_3args_PropertyNamesContainsEmptyTest()
        {
            var dnActual = Directory.GetCurrentDirectory();
            var diActual = new DirectoryInfo(dnActual);
            var propNames = new[] { "Name", "FullName", string.Empty }; // contains string.Empty.
            var propTypes = new[] { typeof(string), typeof(string), typeof(string) };
            Exception exception = null;
            try
            {
                ReflectionAssert.HasProperties(diActual, propNames, propTypes); // propertyNames contains string.Empty
            }
            catch (ReflectionAssertException ex)
            {
                Console.WriteLine(ex.Message);
                exception = ex;
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                throw;
            }
            Assert.IsNotNull(exception);
            Assert.IsInstanceOfType(exception, typeof(ReflectionAssertException));
            StringAssert.Contains(exception.Message, "ReflectionAssert.HasProperties");
            Assert.IsNotNull(exception.InnerException);
            Assert.IsInstanceOfType(exception.InnerException, typeof(ArgumentException));
            StringAssert.Contains(exception.InnerException.Message, new ArgumentException().Message);
            StringAssert.Contains(exception.InnerException.Message, "propertyNames");
        }

        [TestMethod] // do not use ExpectedExceptionAttribute().
        public void HasProperties_3args_PropertyNamesContainsNotExistTest()
        {
            var dnActual = Directory.GetCurrentDirectory();
            var diActual = new DirectoryInfo(dnActual);
            var propNames = new[] { "Name", "FullName", "NoName" }; // contains no exist.
            var propTypes = new[] { typeof(string), typeof(string), typeof(string) };
            Exception exception = null;
            try
            {
                ReflectionAssert.HasProperties(diActual, propNames, propTypes); // contains no exist.
            }
            catch (ReflectionAssertException ex)
            {
                Console.WriteLine(ex.Message);
                exception = ex;
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                throw;
            }
            Assert.IsNotNull(exception);
            Assert.IsInstanceOfType(exception, typeof(ReflectionAssertException));
            StringAssert.Contains(exception.Message, "ReflectionAssert.HasProperties");
            Assert.IsNotNull(exception.InnerException);
            Assert.IsInstanceOfType(exception.InnerException, typeof(ArgumentException));
            //StringAssert.Contains(exception.InnerException.Message, new ArgumentException().Message);
            StringAssert.Contains(exception.InnerException.Message, "NoName");
        }

        [TestMethod] // do not use ExpectedExceptionAttribute().
        public void HasProperties_3args_PropertyNamesTooManyTest()
        {
            var dnActual = Directory.GetCurrentDirectory();
            var diActual = new DirectoryInfo(dnActual);
            var propNames = new[] { "Name", "FullName", "Exists" }; // too many
            var propTypes = new[] { typeof(string), typeof(string) };
            Exception exception = null;
            try
            {
                ReflectionAssert.HasProperties(diActual, propNames, propTypes); // propertyNames has too many items
            }
            catch (ReflectionAssertException ex)
            {
                Console.WriteLine(ex.Message);
                exception = ex;
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                throw;
            }
            Assert.IsNotNull(exception);
            Assert.IsInstanceOfType(exception, typeof(ReflectionAssertException));
            StringAssert.Contains(exception.Message, "ReflectionAssert.HasProperties");
            Assert.IsNotNull(exception.InnerException);
            Assert.IsInstanceOfType(exception.InnerException, typeof(ArgumentOutOfRangeException));
            StringAssert.Contains(exception.InnerException.Message, new ArgumentOutOfRangeException().Message);
            StringAssert.Contains(exception.InnerException.Message, "propertyNames");
        }

        [TestMethod] // do not use ExpectedExceptionAttribute().
        public void HasProperties_3args_PropertyTypesNullTest()
        {
            var dnActual = Directory.GetCurrentDirectory();
            var diActual = new DirectoryInfo(dnActual);
            var propNames = new[] { "Name", "FullName" }; // exist.
            var propTypes = new[] { typeof(string), typeof(string) };
            Exception exception = null;
            try
            {
                ReflectionAssert.HasProperties(diActual, propNames, (IEnumerable<Type>)null); // propertyTypes = null
            }
            catch (ReflectionAssertException ex)
            {
                Console.WriteLine(ex.Message);
                exception = ex;
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                throw;
            }
            Assert.IsNotNull(exception);
            Assert.IsInstanceOfType(exception, typeof(ReflectionAssertException));
            StringAssert.Contains(exception.Message, "ReflectionAssert.HasProperties");
            Assert.IsNotNull(exception.InnerException);
            Assert.IsInstanceOfType(exception.InnerException, typeof(ArgumentNullException));
            StringAssert.Contains(exception.InnerException.Message, new ArgumentNullException().Message);
            StringAssert.Contains(exception.InnerException.Message, "propertyTypes");
        }

        [TestMethod] // do not use ExpectedExceptionAttribute().
        public void HasProperties_3args_PropertyTypesContainsNullTest()
        {
            var dnActual = Directory.GetCurrentDirectory();
            var diActual = new DirectoryInfo(dnActual);
            var propNames = new[] { "Name", "FullName", "Exists" };
            var propTypes = new[] { typeof(string), typeof(string), null }; // contains null.
            Exception exception = null;
            try
            {
                ReflectionAssert.HasProperties(diActual, propNames, propTypes); // propertyTypes contains null
            }
            catch (ReflectionAssertException ex)
            {
                Console.WriteLine(ex.Message);
                exception = ex;
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                throw;
            }
            Assert.IsNotNull(exception);
            Assert.IsInstanceOfType(exception, typeof(ReflectionAssertException));
            StringAssert.Contains(exception.Message, "ReflectionAssert.HasProperties");
            Assert.IsNotNull(exception.InnerException);
            Assert.IsInstanceOfType(exception.InnerException, typeof(ArgumentNullException));
            StringAssert.Contains(exception.InnerException.Message, new ArgumentNullException().Message);
            StringAssert.Contains(exception.InnerException.Message, "propertyTypes");
        }

        [TestMethod] // do not use ExpectedExceptionAttribute().
        public void HasProperties_3args_PropertyTypesContainsMismatchTypeTest()
        {
            var dnActual = Directory.GetCurrentDirectory();
            var diActual = new DirectoryInfo(dnActual);
            var propNames = new[] { "Name", "FullName", "Exists" };
            var propTypes = new[] { typeof(string), typeof(string), typeof(string) }; // contains mismatch type
            Exception exception = null;
            try
            {
                ReflectionAssert.HasProperties(diActual, propNames, propTypes); // propertyTypes contains mismatch type
            }
            catch (ReflectionAssertException ex)
            {
                Console.WriteLine(ex.Message);
                exception = ex;
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                throw;
            }
            Assert.IsNotNull(exception);
            Assert.IsInstanceOfType(exception, typeof(ReflectionAssertException));
            StringAssert.Contains(exception.Message, "ReflectionAssert.HasProperties");
            Assert.IsNotNull(exception.InnerException);
            Assert.IsInstanceOfType(exception.InnerException, typeof(ArgumentException));
            //StringAssert.Contains(exception.InnerException.Message, new ArgumentException().Message);
            StringAssert.Contains(exception.InnerException.Message, "propertyTypes");
        }

        [TestMethod] // do not use ExpectedExceptionAttribute().
        public void HasProperties_3args_PropertyTypesTooManyTest()
        {
            var dnActual = Directory.GetCurrentDirectory();
            var diActual = new DirectoryInfo(dnActual);
            var propNames = new[] { "Name", "FullName" };
            var propTypes = new[] { typeof(string), typeof(string), typeof(string) }; // too many
            Exception exception = null;
            try
            {
                ReflectionAssert.HasProperties(diActual, propNames, propTypes); // propTypes has too many items
            }
            catch (ReflectionAssertException ex)
            {
                Console.WriteLine(ex.Message);
                exception = ex;
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                throw;
            }
            Assert.IsNotNull(exception);
            Assert.IsInstanceOfType(exception, typeof(ReflectionAssertException));
            StringAssert.Contains(exception.Message, "ReflectionAssert.HasProperties");
            Assert.IsNotNull(exception.InnerException);
            Assert.IsInstanceOfType(exception.InnerException, typeof(ArgumentOutOfRangeException));
            StringAssert.Contains(exception.InnerException.Message, new ArgumentOutOfRangeException().Message);
            StringAssert.Contains(exception.InnerException.Message, "propertyTypes");
        }

        [TestMethod]
        public void HasProperties_3args_msg_Test()
        {
            var dnActual = Directory.GetCurrentDirectory();
            var diActual = new DirectoryInfo(dnActual);
            var propNames = new[] { "Name", "FullName" }; // exist.
            var propTypes = new[] { typeof(string), typeof(string) };
            var message = "message";
            ReflectionAssert.HasProperties(diActual, propNames, message);
        }

        [TestMethod] // do not use ExpectedExceptionAttribute().
        public void HasProperties_3args_msg_ActualNullTest()
        {
            var dnActual = Directory.GetCurrentDirectory();
            var diActual = new DirectoryInfo(dnActual);
            var propNames = new[] { "Name", "FullName" }; // exist.
            var propTypes = new[] { typeof(string), typeof(string) };
            var message = "message";
            Exception exception = null;
            try
            {
                ReflectionAssert.HasProperties(null, propNames, propTypes, message); // actual = null
            }
            catch (ReflectionAssertException ex)
            {
                Console.WriteLine(ex.Message);
                exception = ex;
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                throw;
            }
            Assert.IsNotNull(exception);
            Assert.IsInstanceOfType(exception, typeof(ReflectionAssertException));
            StringAssert.Contains(exception.Message, "ReflectionAssert.HasProperties");
            StringAssert.Contains(exception.Message, message);
            Assert.IsNotNull(exception.InnerException);
            Assert.IsInstanceOfType(exception.InnerException, typeof(ArgumentNullException));
            StringAssert.Contains(exception.InnerException.Message, new ArgumentNullException().Message);
            StringAssert.Contains(exception.InnerException.Message, "actual");
        }

        [TestMethod] // do not use ExpectedExceptionAttribute().
        public void HasProperties_3args_msg_PropertyNamesNullTest()
        {
            var dnActual = Directory.GetCurrentDirectory();
            var diActual = new DirectoryInfo(dnActual);
            var propNames = new[] { "Name", "FullName" }; // exist.
            var propTypes = new[] { typeof(string), typeof(string) };
            var message = "message";
            Exception exception = null;
            try
            {
                ReflectionAssert.HasProperties(diActual, (IEnumerable<string>)null, propTypes, message); // propertyNames = null
            }
            catch (ReflectionAssertException ex)
            {
                Console.WriteLine(ex.Message);
                exception = ex;
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                throw;
            }
            Assert.IsNotNull(exception);
            Assert.IsInstanceOfType(exception, typeof(ReflectionAssertException));
            StringAssert.Contains(exception.Message, "ReflectionAssert.HasProperties");
            StringAssert.Contains(exception.Message, message);
            Assert.IsNotNull(exception.InnerException);
            Assert.IsInstanceOfType(exception.InnerException, typeof(ArgumentNullException));
            StringAssert.Contains(exception.InnerException.Message, new ArgumentNullException().Message);
            StringAssert.Contains(exception.InnerException.Message, "propertyNames");
        }

        [TestMethod] // do not use ExpectedExceptionAttribute().
        public void HasProperties_3args_msg_PropertyNamesContainsNullTest()
        {
            var dnActual = Directory.GetCurrentDirectory();
            var diActual = new DirectoryInfo(dnActual);
            var propNames = new[] { "Name", "FullName", null }; // contains null.
            var propTypes = new[] { typeof(string), typeof(string), typeof(string) };
            var message = "message";
            Exception exception = null;
            try
            {
                ReflectionAssert.HasProperties(diActual, propNames, propTypes, message); // propertyNames contains null
            }
            catch (ReflectionAssertException ex)
            {
                Console.WriteLine(ex.Message);
                exception = ex;
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                throw;
            }
            Assert.IsNotNull(exception);
            Assert.IsInstanceOfType(exception, typeof(ReflectionAssertException));
            StringAssert.Contains(exception.Message, "ReflectionAssert.HasProperties");
            StringAssert.Contains(exception.Message, message);
            Assert.IsNotNull(exception.InnerException);
            Assert.IsInstanceOfType(exception.InnerException, typeof(ArgumentNullException));
            StringAssert.Contains(exception.InnerException.Message, new ArgumentNullException().Message);
            StringAssert.Contains(exception.InnerException.Message, "propertyNames");
        }

        [TestMethod] // do not use ExpectedExceptionAttribute().
        public void HasProperties_3args_msg_PropertyNamesContainsEmptyTest()
        {
            var dnActual = Directory.GetCurrentDirectory();
            var diActual = new DirectoryInfo(dnActual);
            var propNames = new[] { "Name", "FullName", string.Empty }; // contains string.Empty.
            var propTypes = new[] { typeof(string), typeof(string), typeof(string) };
            var message = "message";
            Exception exception = null;
            try
            {
                ReflectionAssert.HasProperties(diActual, propNames, propTypes, message); // propertyNames contains string.Empty
            }
            catch (ReflectionAssertException ex)
            {
                Console.WriteLine(ex.Message);
                exception = ex;
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                throw;
            }
            Assert.IsNotNull(exception);
            Assert.IsInstanceOfType(exception, typeof(ReflectionAssertException));
            StringAssert.Contains(exception.Message, "ReflectionAssert.HasProperties");
            StringAssert.Contains(exception.Message, message);
            Assert.IsNotNull(exception.InnerException);
            Assert.IsInstanceOfType(exception.InnerException, typeof(ArgumentException));
            StringAssert.Contains(exception.InnerException.Message, new ArgumentException().Message);
            StringAssert.Contains(exception.InnerException.Message, "propertyNames");
        }

        [TestMethod] // do not use ExpectedExceptionAttribute().
        public void HasProperties_3args_msg_PropertyNamesContainsNotExistTest()
        {
            var dnActual = Directory.GetCurrentDirectory();
            var diActual = new DirectoryInfo(dnActual);
            var propNames = new[] { "Name", "FullName", "NoName" }; // contains no exist.
            var propTypes = new[] { typeof(string), typeof(string), typeof(string) };
            var message = "message";
            Exception exception = null;
            try
            {
                ReflectionAssert.HasProperties(diActual, propNames, propTypes, message); // contains no exist.
            }
            catch (ReflectionAssertException ex)
            {
                Console.WriteLine(ex.Message);
                exception = ex;
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                throw;
            }
            Assert.IsNotNull(exception);
            Assert.IsInstanceOfType(exception, typeof(ReflectionAssertException));
            StringAssert.Contains(exception.Message, "ReflectionAssert.HasProperties");
            StringAssert.Contains(exception.Message, message);
            Assert.IsNotNull(exception.InnerException);
            Assert.IsInstanceOfType(exception.InnerException, typeof(ArgumentException));
            //StringAssert.Contains(exception.InnerException.Message, new ArgumentException().Message);
            StringAssert.Contains(exception.InnerException.Message, "NoName");
        }

        [TestMethod] // do not use ExpectedExceptionAttribute().
        public void HasProperties_3args_msg_PropertyNamesTooManyTest()
        {
            var dnActual = Directory.GetCurrentDirectory();
            var diActual = new DirectoryInfo(dnActual);
            var propNames = new[] { "Name", "FullName", "Exists" }; // too many
            var propTypes = new[] { typeof(string), typeof(string) };
            var message = "message";
            Exception exception = null;
            try
            {
                ReflectionAssert.HasProperties(diActual, propNames, propTypes, message); // propertyNames has too many items
            }
            catch (ReflectionAssertException ex)
            {
                Console.WriteLine(ex.Message);
                exception = ex;
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                throw;
            }
            Assert.IsNotNull(exception);
            Assert.IsInstanceOfType(exception, typeof(ReflectionAssertException));
            StringAssert.Contains(exception.Message, "ReflectionAssert.HasProperties");
            StringAssert.Contains(exception.Message, message);
            Assert.IsNotNull(exception.InnerException);
            Assert.IsInstanceOfType(exception.InnerException, typeof(ArgumentOutOfRangeException));
            StringAssert.Contains(exception.InnerException.Message, new ArgumentOutOfRangeException().Message);
            StringAssert.Contains(exception.InnerException.Message, "propertyNames");
        }

        [TestMethod] // do not use ExpectedExceptionAttribute().
        public void HasProperties_3args_msg_PropertyTypesNullTest()
        {
            var dnActual = Directory.GetCurrentDirectory();
            var diActual = new DirectoryInfo(dnActual);
            var propNames = new[] { "Name", "FullName" }; // exist.
            var propTypes = new[] { typeof(string), typeof(string) };
            var message = "message";
            Exception exception = null;
            try
            {
                ReflectionAssert.HasProperties(diActual, propNames, (IEnumerable<Type>)null, message); // propertyTypes = null
            }
            catch (ReflectionAssertException ex)
            {
                Console.WriteLine(ex.Message);
                exception = ex;
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                throw;
            }
            Assert.IsNotNull(exception);
            Assert.IsInstanceOfType(exception, typeof(ReflectionAssertException));
            StringAssert.Contains(exception.Message, "ReflectionAssert.HasProperties");
            StringAssert.Contains(exception.Message, message);
            Assert.IsNotNull(exception.InnerException);
            Assert.IsInstanceOfType(exception.InnerException, typeof(ArgumentNullException));
            StringAssert.Contains(exception.InnerException.Message, new ArgumentNullException().Message);
            StringAssert.Contains(exception.InnerException.Message, "propertyTypes");
        }

        [TestMethod] // do not use ExpectedExceptionAttribute().
        public void HasProperties_3args_msg_PropertyTypesContainsNullTest()
        {
            var dnActual = Directory.GetCurrentDirectory();
            var diActual = new DirectoryInfo(dnActual);
            var propNames = new[] { "Name", "FullName", "Exists" };
            var propTypes = new[] { typeof(string), typeof(string), null }; // contains null.
            var message = "message";
            Exception exception = null;
            try
            {
                ReflectionAssert.HasProperties(diActual, propNames, propTypes, message); // propertyTypes contains null
            }
            catch (ReflectionAssertException ex)
            {
                Console.WriteLine(ex.Message);
                exception = ex;
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                throw;
            }
            Assert.IsNotNull(exception);
            Assert.IsInstanceOfType(exception, typeof(ReflectionAssertException));
            StringAssert.Contains(exception.Message, "ReflectionAssert.HasProperties");
            StringAssert.Contains(exception.Message, message);
            Assert.IsNotNull(exception.InnerException);
            Assert.IsInstanceOfType(exception.InnerException, typeof(ArgumentNullException));
            StringAssert.Contains(exception.InnerException.Message, new ArgumentNullException().Message);
            StringAssert.Contains(exception.InnerException.Message, "propertyTypes");
        }

        [TestMethod] // do not use ExpectedExceptionAttribute().
        public void HasProperties_3args_msg_PropertyTypesContainsMismatchTypeTest()
        {
            var dnActual = Directory.GetCurrentDirectory();
            var diActual = new DirectoryInfo(dnActual);
            var propNames = new[] { "Name", "FullName", "Exists" };
            var propTypes = new[] { typeof(string), typeof(string), typeof(string) }; // contains mismatch type
            var message = "message";
            Exception exception = null;
            try
            {
                ReflectionAssert.HasProperties(diActual, propNames, propTypes, message); // propertyTypes contains mismatch type
            }
            catch (ReflectionAssertException ex)
            {
                Console.WriteLine(ex.Message);
                exception = ex;
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                throw;
            }
            Assert.IsNotNull(exception);
            Assert.IsInstanceOfType(exception, typeof(ReflectionAssertException));
            StringAssert.Contains(exception.Message, "ReflectionAssert.HasProperties");
            StringAssert.Contains(exception.Message, message);
            Assert.IsNotNull(exception.InnerException);
            Assert.IsInstanceOfType(exception.InnerException, typeof(ArgumentException));
            //StringAssert.Contains(exception.InnerException.Message, new ArgumentException().Message);
            StringAssert.Contains(exception.InnerException.Message, "propertyTypes");
        }

        [TestMethod] // do not use ExpectedExceptionAttribute().
        public void HasProperties_3args_msg_PropertyTypesTooManyTest()
        {
            var dnActual = Directory.GetCurrentDirectory();
            var diActual = new DirectoryInfo(dnActual);
            var propNames = new[] { "Name", "FullName" };
            var propTypes = new[] { typeof(string), typeof(string), typeof(string) }; // too many
            var message = "message";
            Exception exception = null;
            try
            {
                ReflectionAssert.HasProperties(diActual, propNames, propTypes, message); // propTypes has too many items
            }
            catch (ReflectionAssertException ex)
            {
                Console.WriteLine(ex.Message);
                exception = ex;
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                throw;
            }
            Assert.IsNotNull(exception);
            Assert.IsInstanceOfType(exception, typeof(ReflectionAssertException));
            StringAssert.Contains(exception.Message, "ReflectionAssert.HasProperties");
            StringAssert.Contains(exception.Message, message);
            Assert.IsNotNull(exception.InnerException);
            Assert.IsInstanceOfType(exception.InnerException, typeof(ArgumentOutOfRangeException));
            StringAssert.Contains(exception.InnerException.Message, new ArgumentOutOfRangeException().Message);
            StringAssert.Contains(exception.InnerException.Message, "propertyTypes");
        }
    }
}
